Adding Lessons About TDD Clean Architecture Course (#45)
dwyl/learn-flutter: Flutter_TDD_Architecture_Course_notes.md

![First_Test_Pass](https://user-images.githubusercontent.com/27420533/74528160-6d9fef80-4f1f-11ea-9198-64d1aaefb3c4.png)

> Full code of `get_concrete_number_trivia_test.dart` file.

Reviewer: Miguel, when I wrote that you needed to include the "full code" it meant the code needs to be committed to GitHub and _linked_ to from the Markdown file, it does not mean copy pasting the code (duplicating) into the markdown file. Thankfully you have overcome your obstacle to committing and pushing code to GitHub so you will no longer _need_ to paste the code into markdown files and can instead follow the examples provided where the _link_ to the code snapshot is included at the end of each section.

  Future<bool> get isConnected => connectionChecker.hasConnection;
}
```
#### 8.  Local Data Source
> Video 8: https://www.youtube.com/watch?v=fCguzcvLka8

> In the `eighth` tutorial is implemented the shared_preferences function.

Reviewer: This is a PT sentence structure. The EN version is:
```suggestion
In the `eighth` tutorial the shared_preferences function is implemented.
```

It uses the `shared_preferences` function because we only need to save one given NumberTrivia.

Let's add the code to the `number_trivia_local_data_source_test.dart` file.

```dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/mockito.dart';
import 'package:matcher/matcher.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../../../../fixtures/fixture_reader.dart';

class MockSharedPreferences extends Mock implements SharedPreferences {}

void main() {
  NumberTriviaLocalDataSourceImpl dataSource;
  MockSharedPreferences mockSharedPreferences;

  setUp(() {
    mockSharedPreferences = MockSharedPreferences();
    dataSource = NumberTriviaLocalDataSourceImpl(
      sharedPreferences: mockSharedPreferences,
    );
  });
}
```
Now the file corresponding to the `number_trivia_local_data_source.dart`.

```dart

> ...

Reviewer: did you _mean_ to use the [spread operator](https://dart.dev/guides/language/language-tour#spread-operator) `...` here?
if you didn't mean to have `...` please remove. It will create confusion in the mind of a beginner.
```suggestion

```

class NumberTriviaLocalDataSourceImpl implements NumberTriviaLocalDataSource {
  final SharedPreferences sharedPreferences;

  NumberTriviaLocalDataSourceImpl({@required this.sharedPreferences});

  @override
  Future<NumberTriviaModel> getLastNumberTrivia() {
    // TODO: implement getLastNumberTrivia
    return null;
  }

  @override
  Future<void> cacheNumberTrivia(NumberTriviaModel triviaToCache) {
    // TODO: implement cacheNumberTrivia
    return null;
  }
}
```

> Then we created another `fixture` with the name `trivia_cached.json`. With the code.

Reviewer: ```suggestion
Create another `fixture` called `trivia_cached.json` and paste the following code in it:
```

```dart
{
  "text": "Test Text",
  "number": 1
}
```
#### getLastNumberTrivia

> We now make the first test, with the first feature that will show a number stored in the cache.

Reviewer: This sentence is ordered like it was written in PT and then translated to EN.  It can be modified to:
```suggestion
Create the first test for the first feature that will show a number stored in the cache.
```

>  So let's go to the test file `number_trivia_local_data_source_test.dart`.

Reviewer: ```suggestion
Open the `number_trivia_local_data_source_test.dart` file and add the following test:
```

      // act
      final result = await dataSource.getLastNumberTrivia();
      // assert
      verify(mockSharedPreferences.getString('CACHED_NUMBER_TRIVIA'));
      expect(result, equals(tNumberTriviaModel));
    },
  );
});
```
Then go to the corresponding file `number_trivia_local_data_source.dart`.

```dart
@override
Future<NumberTriviaModel> getLastNumberTrivia() {
  final jsonString = sharedPreferences.getString('CACHED_NUMBER_TRIVIA');
  // Future which is immediately completed
  return Future.value(NumberTriviaModel.fromJson(json.decode(jsonString)));
}
```
We move on to the second test that deals with when there is no number in the cache to show.

```dart
test('should throw a CacheException when there is not a cached value', () {
  // arrange
  when(mockSharedPreferences.getString(any)).thenReturn(null);
  // act
  // Not calling the method here, just storing it inside a call variable
  final call = dataSource.getLastNumberTrivia;
  // assert
  // Calling the method happens from a higher-order function passed.
  // This is needed to test if calling a method throws an exception.
  expect(() => call(), throwsA(TypeMatcher<CacheException>()));
});
```

> Then just implement the function to the file `number_trivia_local_data_source.dart`.

Reviewer: ```suggestion
This test will fail as the `dataSource.getLastNumberTrivia` method is not yet implemented. <br />
Let's implement the function in the `number_trivia_local_data_source.dart` file:
```

@override
Future<NumberTriviaModel> getLastNumberTrivia() {
  final jsonString = sharedPreferences.getString('CACHED_NUMBER_TRIVIA');
  if (jsonString != null) {
    return Future.value(NumberTriviaModel.fromJson(json.decode(jsonString)));
  } else {
    throw CacheException();
  }
}
```
We'll move on to the third test, which will check if any functions have been called.

```dart
group('cacheNumberTrivia', () {
  final tNumberTriviaModel =
      NumberTriviaModel(number: 1, text: 'test trivia');

  test('should call SharedPreferences to cache the data', () {
    // act
    dataSource.cacheNumberTrivia(tNumberTriviaModel);
    // assert
    final expectedJsonString = json.encode(tNumberTriviaModel.toJson());
    verify(mockSharedPreferences.setString(
      CACHED_NUMBER_TRIVIA,
      expectedJsonString,
    ));
  });
});
```
```dart
@override
Future<void> cacheNumberTrivia(NumberTriviaModel triviaToCache) {
  return sharedPreferences.setString(
    CACHED_NUMBER_TRIVIA,
    json.encode(triviaToCache.toJson()),
  );
}
```

#### 9.  Remote Data Source
> Video 9: https://www.youtube.com/watch?v=msGsYPtZnhU

Tutorial number `nine` will handle all communication with the numbers API.

> The first thing to do is create a file corresponding to the `number_trivia_remote_data_source.dart`.

Reviewer: ```suggestion
Create a file called `number_trivia_remote_data_source.dart`.
```

> The file must be created inside the `test` folder.

Reviewer: this is included in the sentence above.
```suggestion

```

> We start by setting up the `test` file.

Reviewer: ```suggestion

```

```dart
import 'package:http/http.dart' as http;
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/mockito.dart';
import 'package:matcher/matcher.dart';

import '../../../../fixtures/fixture_reader.dart';

class MockHttpClient extends Mock implements http.Client {}

void main() {
  NumberTriviaRemoteDataSourceImpl dataSource;
  MockHttpClient mockHttpClient;

  setUp(() {
    mockHttpClient = MockHttpClient();
    dataSource = NumberTriviaRemoteDataSourceImpl(client: mockHttpClient);
  });
}
```

> Then we implement in the file corresponding to the test file, the `number_trivia_remote_data_source.dart`.

Reviewer: ```suggestion
Implement `NumberTriviaRemoteDataSourceImpl` in the `number_trivia_remote_data_source.dart` file:
```

```dart
class NumberTriviaRemoteDataSourceImpl implements NumberTriviaRemoteDataSource {
  final http.Client client;

  NumberTriviaRemoteDataSourceImpl({@required this.client});

  @override
  Future<NumberTriviaModel> getConcreteNumberTrivia(int number) {
    // TODO: implement getConcreteNumberTrivia
    return null;
  }

  @override
  Future<NumberTriviaModel> getRandomNumberTrivia() {
    // TODO: implement getRandomNumberTrivia
    return null;
  }
}
```

> Now we do a test for `GET request on a URL with number` being the endpoint.

Reviewer: ```suggestion
Now create a test for the `GET request on a URL with number` being the endpoint.
```

```dart
group('getConcreteNumberTrivia', () {
  final tNumber = 1;

  test(
    'should preform a GET request on a URL with number being the endpoint and with application/json header',
    () {
      //arrange
      when(mockHttpClient.get(any, headers: anyNamed('headers'))).thenAnswer(
        (_) async => http.Response(fixture('trivia.json'), 200),
      );
      // act
      dataSource.getConcreteNumberTrivia(tNumber);
      // assert
      verify(mockHttpClient.get(
        'http://numbersapi.com/$tNumber',
        headers: {'Content-Type': 'application/json'},
      ));
    },
  );
}
```

As always, it is necessary to `implement` the file corresponding to the test file.

```dart
@override
Future<NumberTriviaModel> getConcreteNumberTrivia(int number) {
  client.get(
    'http://numbersapi.com/$number',
    headers: {'Content-Type': 'application/json'},
  );
}
```

> The next `test` is for when we get a number of the application without errors.

Reviewer: ```suggestion
The next `test` is for getting a number without errors.
```

```dart
final tNumberTriviaModel =
    NumberTriviaModel.fromJson(json.decode(fixture('trivia.json')));
...
test(
  'should return NumberTrivia when the response code is 200 (success)',
  () async {
    // arrange
    when(mockHttpClient.get(any, headers: anyNamed('headers'))).thenAnswer(
      (_) async => http.Response(fixture('trivia.json'), 200),
    );
    // act
    final result = await dataSource.getConcreteNumberTrivia(tNumber);
    // assert
    expect(result, equals(tNumberTriviaModel));
  },
);
```

> Then we do the `implementation`.

Reviewer: ```suggestion
Then we create the `implementation`.
```

```dart
@override
Future<NumberTriviaModel> getConcreteNumberTrivia(int number) async {
  final response = await client.get(
    'http://numbersapi.com/$number',
    headers: {'Content-Type': 'application/json'},
  );

  return NumberTriviaModel.fromJson(json.decode(response.body));
}
```

Now we need to run a `test` for when there is some kind of error.

```dart
test(
  'should throw a ServerException when the response code is 404 or other',
  () async {
    // arrange
    when(mockHttpClient.get(any, headers: anyNamed('headers'))).thenAnswer(
      (_) async => http.Response('Something went wrong', 404),
    );
    // act
    final call = dataSource.getConcreteNumberTrivia;
    // assert
    expect(() => call(tNumber), throwsA(TypeMatcher<ServerException>()));
  },
);
```

> We `implement` in the corresponding file using a if .

Reviewer: Which file does the implementation go in?
```suggestion
Implement the code to make the test pass in the corresponding file using an `if`:
```

        headers: {'Content-Type': 'application/json'},
      ));
    },
  );

  test(
    'should return NumberTrivia when the response code is 200 (success)',
    () async {
      // arrange
      setUpMockHttpClientSuccess200();
      // act
      final result = await dataSource.getRandomNumberTrivia();
      // assert
      expect(result, equals(tNumberTriviaModel));
    },
  );

  test(
    'should throw a ServerException when the response code is 404 or other',
    () async {
      // arrange
      setUpMockHttpClientFailure404();
      // act
      final call = dataSource.getRandomNumberTrivia;
      // assert
      expect(() => call(), throwsA(TypeMatcher<ServerException>()));
    },
  );
});
```

Now we implement the random function in the `number_trivia_remote_data_source.dart` file.

```dart
@override
Future<NumberTriviaModel> getRandomNumberTrivia() async {
  final response = await client.get(
    'http://numbersapi.com/random',
    headers: {'Content-Type': 'application/json'},
  );

  if (response.statusCode == 200) {
    return NumberTriviaModel.fromJson(json.decode(response.body));
  } else {
    throw ServerException();
  }
}
```

> Now the tests will `pass` on both sides.

Reviewer: Please _show_ the tests passing.

Author: I've added the screenshot of all the tests passing.

Reviewer: Thanks. 👍
